### ISSUE TITLE
Host not preserved w/ request_path + preserve_host

### ISSUE THREAD

**Issue by deltheil** [Mashape/kong#1065]
It looks like with Kong 0.6.0 or higher the host is not correctly preserved for API-s configured with:

1. `request_path` routing,
2. `preserve_host=true` option.

I faced such a problem after a 0.5.4 to 0.7.0 migration.  With Kong 0.5.4, the host is obtained via the corresponding [nginx header](https://github.com/Mashape/kong/blob/bdc6043b6295bb252f3ce6e58c8f48e2cf760ad3/kong/resolver/access.lua#L218).

With Kong +0.6.0 the logic evolved:

* [`matched_host` is `nil`](https://github.com/Mashape/kong/blob/4943619e1b68c111caaa11e8c369ee173896f31d/kong/core/resolver.lua#L209) post lookup by request path,
* and so [`upstream_host` is `nil`](https://github.com/Mashape/kong/blob/4943619e1b68c111caaa11e8c369ee173896f31d/kong/core/resolver.lua#L241).

I patched this on my side with the fallback below, but there should be a better / proper solution:

```Lua
if api.preserve_host then
  upstream_host = matched_host or ngx.req.get_headers()["host"]
end
```

**Comment by pklingem** [comment]
seen this as well

**Comment by thefosk** [comment]
Thanks guys, this issue has been closed by #1068.
